chia/wallet/cat_wallet/cat_utils.py

```diff
@@ -30,15 +30,15 @@ class SpendableCAT:
     limitations_program_reveal: Program = Program.to([])
 
 
-def match_cat_puzzle(puzzle: Program) -> Tuple[bool, Iterator[Program]]:
+def match_cat_puzzle(mod: Program, curried_args: Program) -> Optional[Iterator[Program]]:
```

Comment:
this is the core change, along with similar changes to the corresponding function signature in the NFT and DID wallets. i.e. instead of passing in the full puzzle, the underlying program and the curried arguments are passed in.

This lets the 3 different kinds of wallets (CAT, DID and NFT) all use the result from a single call to `uncurry()`, instead of all 3 calling it themselves.